File: templates/components/engines/typescript/agent/tools/duckduckgo.ts
import { JSONSchemaType } from "ajv";
import { ImageSearchOptions, search, searchImages } from "duck-duck-scrape";
import { BaseTool, ToolMetadata } from "llamaindex";

export type DuckDuckGoParameter = {
  query: string;
  region?: string;
  maxResults?: number;
};

export type DuckDuckGoToolParams = {
  metadata?: ToolMetadata<JSONSchemaType<DuckDuckGoParameter>>;
};

const DEFAULT_SEARCH_METADATA: ToolMetadata<
  JSONSchemaType<DuckDuckGoParameter>
> = {
  name: "duckduckgo_search",
  description:
    "Use this function to search for information in the internet using DuckDuckGo.",
  parameters: {
    type: "object",
    properties: {
      query: {
        type: "string",
        description: "The query to search in DuckDuckGo.",
      },
      region: {
        type: "string",
        description:
          "Optional, The region to be used for the search in [country-language] convention, ex us-en, uk-en, ru-ru, etc...",
        nullable: true,
      },
      maxResults: {
        type: "number",
        description:
          "Optional, The maximum number of results to be returned. Default is 10.",
        nullable: true,
      },
    },
    required: ["query"],
  },
};

const DEFAULT_IMAGE_SEARCH_METADATA: ToolMetadata<
  JSONSchemaType<DuckDuckGoParameter>
> = {
  name: "duckduckgo_image_search",
  description: "Use this function to search for images in DuckDuckGo.",
  parameters: {
    type: "object",
    properties: {
      query: {
        type: "string",
        description: "The query to search in DuckDuckGo.",
      },
      region: {
        type: "string",
        description:
          "Optional, The region to be used for the search in [country-language] convention, ex us-en, uk-en, ru-ru, etc...",
        nullable: true,
      },
      maxResults: {
        type: "number",
        description:
          "Optional, The maximum number of results to be returned. Default is 10.",
        nullable: true,
      },
    },
    required: ["query"],
  },
};

type DuckDuckGoSearchResult = {
  title: string;
  description: string;
  url: string;
};

type DuckDuckGoImageResult = {
  image: string;
  title: string;
  source: string;
  url: string;
};

export class DuckDuckGoSearchTool implements BaseTool<DuckDuckGoParameter> {
  metadata: ToolMetadata<JSONSchemaType<DuckDuckGoParameter>>;

  constructor(params: DuckDuckGoToolParams) {
    this.metadata = params.metadata ?? DEFAULT_SEARCH_METADATA;
  }

  async call(input: DuckDuckGoParameter) {
    const { query, region, maxResults = 10 } = input;
    const options = region ? { region } : {};
    // Temporarily sleep to reduce overloading the DuckDuckGo
    await new Promise((resolve) => setTimeout(resolve, 1000));
    const searchResults = await search(query, options);

    return searchResults.results.slice(0, maxResults).map((result) => {
      return {
        title: result.title,
        description: result.description,
        url: result.url,
      } as DuckDuckGoSearchResult;
    });
  }
}

export class DuckDuckGoImageSearchTool
  implements BaseTool<DuckDuckGoParameter>
{
  metadata: ToolMetadata<JSONSchemaType<DuckDuckGoParameter>>;

  constructor(params: DuckDuckGoToolParams) {
    this.metadata = params.metadata ?? DEFAULT_IMAGE_SEARCH_METADATA;
  }

  async call(input: DuckDuckGoParameter) {
    const { query, region, maxResults = 5 } = input;
    const options: Partial<ImageSearchOptions> = region
      ? { locale: region }
      : {};
    // Temporarily sleep to reduce overloading the DuckDuckGo
    await new Promise((resolve) => setTimeout(resolve, 1000));
    const imageResults = await searchImages(query, options);

    return imageResults.results.slice(0, maxResults).map((result) => {
      return {
        image: result.image,
        title: result.title,
        source: result.source,
        url: result.url,
      } as DuckDuckGoImageResult;
    });
  }
}

export function getTools() {
  return [new DuckDuckGoSearchTool({}), new DuckDuckGoImageSearchTool({})];
}

File: templates/types/streaming/fastapi/app/engine/tools/duckduckgo.py
from llama_index.core.tools.function_tool import FunctionTool


def duckduckgo_search(
    query: str,
    region: str = "wt-wt",
    max_results: int = 10,
):
    """
    Use this function to search for any query in DuckDuckGo.
    Args:
        query (str): The query to search in DuckDuckGo.
        region Optional(str): The region to be used for the search in [country-language] convention, ex us-en, uk-en, ru-ru, etc...
        max_results Optional(int): The maximum number of results to be returned. Default is 10.
    """
    try:
        from duckduckgo_search import DDGS
    except ImportError:
        raise ImportError(
            "duckduckgo_search package is required to use this function."
            "Please install it by running: `poetry add duckduckgo_search` or `pip install duckduckgo_search`"
        )

    results = []
    with DDGS() as ddg:
        results = list(
            ddg.text(
                keywords=query,
                region=region,
                max_results=max_results,
            )
        )
    return results


def duckduckgo_image_search(
    query: str,
    region: str = "wt-wt",
    max_results: int = 10,
):
    """
    Use this function to search for images in DuckDuckGo.
    Args:
        query (str): The query to search in DuckDuckGo.
        region Optional(str): The region to be used for the search in [country-language] convention, ex us-en, uk-en, ru-ru, etc...
        max_results Optional(int): The maximum number of results to be returned. Default is 10.
    """
    try:
        from duckduckgo_search import DDGS
    except ImportError:
        raise ImportError(
            "duckduckgo_search package is required to use this function."
            "Please install it by running: `poetry add duckduckgo_search` or `pip install duckduckgo_search`"
        )
    with DDGS() as ddg:
        results = list(
            ddg.images(
                keywords=query,
                region=region,
                max_results=max_results,
            )
        )
    return results


def get_tools(**kwargs):
    return [
        FunctionTool.from_defaults(duckduckgo_search),
        FunctionTool.from_defaults(duckduckgo_image_search),
    ]
